src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -154,6 +154,10 @@ public bool SetNullableContext(byte? value)
         private readonly NamespaceOrTypeSymbol _containingSymbol;
         protected readonly MergedTypeDeclaration declaration;
 
+        // To compute explicitly declared members, binding must be limited (to avoid race conditions where binder cache captures symbols that aren't part of the final set)
+        // This builder should not be modified once saved here
+        private MembersAndInitializersBuilder? _lazyDeclaredMembersAndInitializers;
```

Comment:
>MembersAndInitializersBuilder [](start = 16, length = 29)

I think that things we cache should be immutable. I.e. this shouldn't be a builder, but rather should be an immutable ```MembersAndInitializers```. When we build synthesized members, we probably shouldn't eagerly clone content of this immutable instance. Instead we should use it as an input, figure out if there is anything to add. Create a new immutable instance only when necessary, reusing existing immutable arrays when appropriate (for example static initializers, etc.) 
This should also address the problem with releasing array builders back to the pool.

Reply:
Also, I think we should consider finding a way to "release" this instance once ```_lazyMembersAndInitializers``` is set. Perhaps with some sort of Read/Write lock, where the last thread still using the instance will be able to upgrade read lock to write lock and will safely reset the field to null. 

---
In reply to: [562258069](https://github.com/dotnet/roslyn/pull/50650#discussion_r562258069) [](ancestors = 562258069)

Reply:
I considered that and initially prototyped it, but that means we'll have 4 types (two builders and two immutable structures) and we'll allocate even more to convert from/to different representations. I'd like to hear what others think on this choice.

The part about cloning as lazily as possible I agree with (if we have an immutable result for first result).

---
In reply to: [562261284](https://github.com/dotnet/roslyn/pull/50650#discussion_r562261284) [](ancestors = 562261284,562258069)

Reply:
I agree, we should cache immutable objects rather than builders. If we're caching and sharing builders across threads, when are the builders frozen, and are they thread-safe in the meantime?

---
In reply to: [562310092](https://github.com/dotnet/roslyn/pull/50650#discussion_r562310092) [](ancestors = 562310092,562261284,562258069)